### ISSUE TITLE
[in_app_purchase][android] Use new exposed fields for Billing API V5 migration

### ISSUE THREAD

**Issue by martijnjanssen** [flutter/flutter#107370]
Now that https://github.com/flutter/flutter/issues/88664 has been merged and published, the newly exposed `skus` field should be used to expose available products. This has a deadline of august 2nd due to the deprecation for new apps: [Sell subscriptions](https://developer.android.com/google/play/billing/subscriptions)

**Comment by danagbemava-nc** [comment]
Hi @martijnjanssen, upon further investigation, it looks like the `skus` field has been added. See https://pub.dev/packages/in_app_purchase_android/changelog#023. 

Would that be what you were referring to?

**Comment by martijnjanssen** [comment]
No, that is for the stand-alone android integration. What I'm referring to is the fact that the `in_app_purchase` package does not yet expose these separate products that are now shown in `in_app_purchase_android`.

I'm not sure what the blocking part is, but the latest [in_app_purchase](https://pub.dev/packages/in_app_purchase) release is from before that fix. For someone to fully utilize the new billing api, these separate `skus` need to be returned from the `in_app_purchase` package.

I think that is the issue, if my understanding of the issue is incorrect, please let me know :)

**Comment by danagbemava-nc** [comment]
Hi @martijnjanssen, from what I understand, (for this particular update), it isn't necessary for the in_app_purchase plugin to be updated since this change was done specifically for the in_app_purchase_android plugin.

If you run `flutter pub get`, your project should pull in `in_app_purchase_android: 0.2.3+1`. 

Also, from what I can see, all the logic is being handled in the `in_app_purchase_android` plugin, so nothing else needs to be done in the main `in_app_purchase` plugin. 

Can you confirm if you get the `in_app_purchase_android: 0.2.3+1` version in your project after updating your dependencies?

**Comment by GaryQian** [comment]
cc @GaryQian (for reference)

**Comment by slimbhiri** [comment]
I just started In App Purchases on Google Pay with this Flutter package and I succeed to get `ProductDetails` but as you know Google Play is now managing **Base plans** & **Offers** inside each subscription which I cannot find under `ProductDetail`.

**Source**: https://support.google.com/googleplay/android-developer/answer/12124625

**Comment by yunchiri** [comment]
https://github.com/dooboolab/flutter_inapp_purchase/issues/381#issue-1334317690

it also same problem like this issue 
is this not solved yet right?

**Comment by dhanushka1999** [comment]
Any updates?

**Comment by victormanuelfrancodev** [comment]
I have this problem I can't upload my app, anyone with the same problem?

**Comment by Afsal102** [comment]
@victormanuelfrancodev can update the transistive dependencies? in my case i updated the in_app_purchase_android plugin, and invalidated the gradle  caches and published a new version, that resolved the the issue `Please upgrade to Billing Library version 4 or newer to publish this app.`

**Comment by victormanuelfrancodev** [comment]
@Afsal102  thank you!, worked

**Comment by giorgio79** [comment]
No candy as the main in_app_purchase pubspec.yaml uses an old version. Pls update that https://github.com/flutter/plugins/blob/main/packages/in_app_purchase/in_app_purchase/pubspec.yaml

also discussed at https://github.com/flutter/flutter/issues/111261

**Comment by RamsayGit** [comment]
I'm on the latest version of flutter and the in app purchase plugin and I can only get the first monthly base plan of my google play subscription, I have a second yearly base plan listed and activated in the subscription but I cannot retrieve it with in_app_purchase 3.0.7. Any solution available to this problem or what can I do?

**Comment by kherel** [comment]
> I'm on the latest version of flutter and the in app purchase plugin and I can only get the first monthly base plan of my google play subscription, I have a second yearly base plan listed and activated in the subscription but I cannot retrieve it with in_app_purchase 3.0.7. Any solution available to this problem or what can I do?

remove `pubspec.lock`
and run `flutter clean` and `flutter pub get`

**Comment by Nihildum** [comment]
I have the same problem as @RamsayGit . I did a flutter upgrade and a flutter pub upgrade in my project. I am on in_app_purchase 3.0.8. (Same problem also after what @kherel proposed.)

Is there any additional API call that needs to be done? From the skuDetails in the objects I can not figure out, what I need to do.

**Comment by kinex** [comment]
I think the problem is that the plugin uses the deprecated API `querySkuDetailsAsync` [here](https://github.com/flutter/plugins/blob/main/packages/in_app_purchase/in_app_purchase_android/android/src/main/java/io/flutter/plugins/inapppurchase/MethodCallHandlerImpl.java#L197) . It should use the new API `queryProductDetailsAsync` instead as documented [here](https://developer.android.com/reference/com/android/billingclient/api/BillingClient#querySkuDetailsAsync(com.android.billingclient.api.SkuDetailsParams,%20com.android.billingclient.api.SkuDetailsResponseListener)). This new API allows to get all base plans for a product, which is required to support properly for example monthly and yearly plans following the current best practices. This is a major issue now for example in my app. I can offer only monthly subscriptions in Android.

There seems to be already to-do in the plugin code referencing this [migration guide](https://developer.android.com/google/play/billing/migrate-gpblv5).

**Comment by jimmyff** [comment]
I think @kinex has hit the nail on the head, the `InAppPurchase.instance.queryProductDetails()` will need modifying to query the new subscription format via the [queryProductDetailsAsync](https://developer.android.com/reference/com/android/billingclient/api/BillingClient#queryProductDetailsAsync) method. 

How this will work with things like the offers and there eligibility criteria etc? (see: [Recent changes to subscriptions in Play Console](https://support.google.com/googleplay/android-developer/answer/12124625))

**Comment by seunghwanly** [comment]
Besides the `skus`, I'm willing to use the property called `"name"` in `ProductDetails` since Google's Billing library provides the [`getName` method](https://developer.android.com/google/play/billing/release-notes#4-1) in `BillingClient` class.

👉 [Implemented Code in `in_app_purchase` package](https://github.com/flutter/plugins/blob/bdf1a98a12aaad6887ca73dfd738f90c030f57d2/packages/in_app_purchase/in_app_purchase_android/android/src/main/java/io/flutter/plugins/inapppurchase/Translator.java#L21-L41)

is the `name` property excluded because it would be a stand-alone implementation? I found that StoreKit does not provide a `ProductDetails` with `name + app_name` with a property called `title`.

or else I am looking forward to taking a PR for adding `name` property inside the `ProductDetails`, then `ProductDetails` may have 2 properties called `name` and `title`.

expected results can be like below:
``` dart
name = '${Product Name}'
title = '${Product Name}' + '(' + '${Application Name}' + ')'
```

**Comment by seanharding** [comment]
So currently the only solution here is to create a separate top-level Subscription for each type/billing period, rather than using Base Plans under a single Subscription?

Is anyone actively working on updating the package to handle the new structure properly?

**Comment by AlexanderThiele** [comment]
I'm also waiting for the update. Does someone know if the plugin team is working on it? Is there a design doc to combine the new Google approach with the classic apple approach?

**Comment by jimmyff** [comment]
> So currently the only solution here is to create a separate top-level Subscription for each type/billing period, rather than using Base Plans under a single Subscription?

Yeah that's how I'm doing it at the moment until the flutter package catches up.

> Does someone know if the plugin team is working on it?

@danagbemava-nc any chance of an update on this?

**Comment by danagbemava-nc** [comment]
Hi @jimmyff, unfortunately, that is not information I have

**Comment by Tobias-Georg** [comment]
I have the same problem.

@danagbemava-nc or anybody else: how can we find out if updating the flutter package to properly work with base plans is on the backlog?

**Comment by AlexanderThiele** [comment]
@Tobias-Georg it is probably very difficult to find a similar data object for Apple Store and Google Play. i tried it myself and was not really sure how to handle it correctly for both platforms.

**Comment by Tobias-Georg** [comment]
@AlexanderThiele I agree, it requires a lot of conceptual work to integrate specifics of the platforms into this cross-platform tool. But I hope the team only would give up on it after looking into the challenges in some detail.
It is similar to upgrading/downgrading, which has been solved quite differently in the platforms and is possible nevertheless, right?

**Comment by seanharding** [comment]
I mean, yes, doing things in a cross-platform way is challenging, but that's the precise problem that Flutter claims to solve. I don't see how Flutter can be a first-class mobile development platform if it can't support in-app purchases fully (which is currently the case on multiple levels; I feel like I'm building a little collection of github issues I'm following for the various IAP problems I've run into).

**Comment by ilhanoguzhan** [comment]
The package does not work on android for new apps that google restricts from deprecated calls by `skus` instead of `products` and `product ids` (old product ids has backward compatibility) I had to fork the android package to make it compatible with v5. It's working now for basic subscriptions. If anybody needs quick workaround until official contributors having update on the android package, here is the way:

Add this in pubspec.yaml with root indentation:
```
dependency_overrides:
  in_app_purchase_android:
    git:
      url: git@github.com:ilhanoguzhan/flutter_plugins.git
      path: packages/in_app_purchase/in_app_purchase_android
```

I don't provide any guarantee to work it properly with all functions the the package offers. Feel free to create a fork update more and let me know if have improvement on it.

**Comment by bestdpf** [comment]
> dependency_overrides:
>   in_app_purchase_android:
>     git:
>       url: git@github.com:ilhanoguzhan/flutter_plugins.git
>       path: packages/in_app_purchase/in_app_purchase_android
this way is right, though there are many other works to do. This code is not enough.
